ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java

```diff
@@ -0,0 +1,336 @@
+         * e.g. Merge stmt, multi-insert stmt when mixing DP and SP writes
+         * Insert ... Select ... Union All Select ... using
+         * {@link org.apache.hadoop.hive.ql.exec.AbstractFileMergeOperator#UNION_SUDBIR_PREFIX}
+         */
+        acidSink.setStatementId(driverContext.getTxnManager().getStmtIdAndIncrement());
+        String unionAllSubdir = "/" + AbstractFileMergeOperator.UNION_SUDBIR_PREFIX;
+        if (acidSink.getInsertOverwrite() && acidSink.getDirName().toString().contains(unionAllSubdir) &&
+            acidSink.isFullAcidTable()) {
+          throw new UnsupportedOperationException("QueryId=" + driverContext.getPlan().getQueryId() +
+              " is not supported due to OVERWRITE and UNION ALL.  Please use truncate + insert");
+        }
+      }
+    }
+  }
+
+  private void allocateWriteIdForAcidAnalyzeTable() throws LockException {
+    if (driverContext.getPlan().getAcidAnalyzeTable() != null) {
+      Table table = driverContext.getPlan().getAcidAnalyzeTable().getTable();
+      driverContext.getTxnManager().getTableWriteId(table.getDbName(), table.getTableName());
+    }
+  }
+
+  private boolean setWriteIdForAcidDdl() throws SemanticException, LockException {
+    DDLDescWithWriteId acidDdlDesc = driverContext.getPlan().getAcidDdlDesc();
+    boolean hasAcidDdl = acidDdlDesc != null && acidDdlDesc.mayNeedWriteId();
+    if (hasAcidDdl) {
+      String fqTableName = acidDdlDesc.getFullTableName();
+      TableName tableName = HiveTableName.of(fqTableName);
+      long writeId = driverContext.getTxnManager().getTableWriteId(tableName.getDb(), tableName.getTable());
+      acidDdlDesc.setWriteId(writeId);
+    }
+    return hasAcidDdl;
+  }
+
+  private void acquireLocksInternal() throws CommandProcessorException, LockException {
+    /* It's imperative that {@code acquireLocks()} is called for all commands so that
+       HiveTxnManager can transition its state machine correctly */
+    String userFromUGI = DriverUtils.getUserFromUGI(driverContext);
+    driverContext.getTxnManager().acquireLocks(driverContext.getPlan(), context, userFromUGI, driverState);
+    List<HiveLock> locks = context.getHiveLocks();
+    LOG.info("Operation {} obtained {} locks", driverContext.getPlan().getOperation(),
+        ((locks == null) ? 0 : locks.size()));
+    // This check is for controlling the correctness of the current state
+    if (driverContext.getTxnManager().recordSnapshot(driverContext.getPlan()) &&
+        !driverContext.isValidTxnListsGenerated()) {
+      throw new IllegalStateException("Need to record valid WriteID list but there is no valid TxnID list (" +
+          JavaUtils.txnIdToString(driverContext.getTxnManager().getCurrentTxnId()) +
+          ", queryId: " + driverContext.getPlan().getQueryId() + ")");
+    }
+  }
+
+  public void addHiveLocksFromContext() {
+    hiveLocks.addAll(context.getHiveLocks());
+  }
+
+  public boolean hasHiveLock() {
+    return hiveLocks.isEmpty();
+  }
```

Comment:
I would prefer keeping the locks local to DriverTxnManager, and its' own context(?), and not storing in the global context?

Reply:
I agree, that HiveLocks should be handled together. Currently MoveTask also needs to access them, as it wants to remove somme locks, so to have the hiveLocks only locally in DriverTxnHandler should somehow solve that. I suggest to solve this issue in a separate jira, as this one is only for moving code to a cleaner structure.